### PR TITLE
Aludel hardhat tests

### PR DESCRIPTION
**Pull request by itirabasso**
These are adapted tests from Ampleforth's Geyser repo.

AludelV1 tests pass
AludelV2 tests need a little bit more of work because the `AludelFactory.launch` registers the VaultFactory during the program creation, which breaks some assumptions of the tests.

Merge this after #28

### CODE REVIEW

File: src/contracts/mocks/MockStakeHelper.sol [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+contract MockStakeHelper {
+    function flashStake(
+        address geyser,
+        address vault,
+        uint256 amount,
+        bytes calldata lockPermission,
+        bytes calldata unstakePermission
+    ) external {
+        Aludel(geyser).stake(vault, amount, lockPermission);
```
Comment (plaintextpaco) [inline src/contracts/mocks/MockStakeHelper.sol:18]: I don't know about this particular usecase, but shouldn't there be a hook to do something in between the stake and unstake?

File: test/index.ts [hunk 2]
```diff
@@ -6,17 +6,22 @@ import {
   deployments as hardhatDeployments,
   network,
   ethers as hardhatEthers,
+  getNamedAccounts,
```
Comment (plaintextpaco) [inline test/index.ts:9]: this isn't used

File: test/index.ts [hunk 3]
```diff
@@ -6,17 +6,22 @@ import {
...
 } from "hardhat";
 import{
   Aludel,
   AludelFactory,
+  AludelV2,
```
Comment (plaintextpaco) [inline test/index.ts:14]: if the idea is to only use AludelV2, keep in mind the old Aludel is still being used for the one launched by the factory, line 98

**Review by plaintextpaco — CHANGES_REQUESTED** [review]
commits
-------
- [ ] 29e2a88 rename Aludel.sol to AludelV2.sol
    - [ ] this commit doesn't rename the Aludel, it just uses AludelV2 instead of Aludel in the hardhat tests.
    - [ ] Also, the most important thing in this commit IMO is deploying a new CrucibleFactory in the tests instead of relying on having one under `deployments/`. I think this is cool, however, I think the preferred approach would be to create deploy scripts for the CrucibleFactory and use `deployments.get` in the tests.
- [ ] 9aa237b import Aludel from alchemist repo
    - [ ] what's the purpose of the mocks/imports.sol file? it's not evident so we should at least put it in a comment in it.
- [ ] 655665c use OZ ERC20 implementation for MockERC20
    - [ ] I saw rari-capital's repo was reprecated. Is it related to that? or to something else? the commit message is a good place to put those kind of explainations
- [ ] 5024df1 Update AludelV2 deployment script to use the correct filename
    - [ ] if we're deploying the wrong contract, then that's something worth commenting on the deploy scripts PR instead of silently adding it in an unrelated one. Good catch regardless.
    - [ ] also, the referenced contract isn't there until two commits later, try to be consistent with that
- [ ] 03e65c8 make PowerSwitchFactory deployment script use latest version of the contract
    - [ ] again, worth commenting on the previous PR. Are there several versions of the PowerSwitchFactory? I wasn't aware of that
- [ ] 8f85410 rename Aludel to AludelV2
    - [ ] this doesn't rename the aludel but instead just adds the new contract. Is the old one still necessary?
- [x] 8f6592b add shared-before-each plugin
    - [ ] nice, will see it in the diff along with the new tests
    - [ ] are we sure we don't have to add anything to hardhat.config.ts or package.json?
- [x] ab7c872 add util functions for alchemist test suite
    - [ ] nice, will see it in the diff along with the new tests
    - [ ] are they all used?
    - [ ] we should probably delete the old repeated code throughout the codebase
- [x] 1ced441 add MockStakeHelper contract
- [ ] 7487f8e Add AludelV1 tests
- [ ] 0751b62 (HEAD -> aludel-tests, origin/aludel-tests) add AludelV2 tests

comments
--------
- [ ] I think it's suboptimal to create commits where, for example, a file is added but isn't used anywhere. It'd be better to group in a commit the creation of a file like test/setup.ts together with it's use in existing testsuites and removal of duplicated code. So I, as a reviewer, can check that everything was cleanly removed in that commit and move one to analyze another aspect of the PR, instead of having to go back at it when reviewing the entire diff
- [ ] I'm getting multiple very similar errors like this one, for both the AludelV2 and the Geyser tests:

    test/Geyser.ts:2331:31 - error TS2345: Argument of type 'Wallet' is not assignable to parameter of type 'string | Signer | Provider'.

    2331           await vault.connect(user).rageQuit(geyser.address, stakingToken.address, {
                                       ^^^^

I've:
- installed dependencies: `yarn`
- generated typechain bindings: `yarn hardhat typechain`
- compiled the project: `yarn hardhat compile`
- run the tests: `yarn hardhat test` -- and got the error in this step

I don't know if this is because the tests are a WIP or because you forgot to commit something, but I comment this just in case. I didn't review the tests themselves yet for this reason.

File: src/contracts/mocks/MockStakeHelper.sol (hunk 1 above)
Comment (itirabasso) [reply src/contracts/mocks/MockStakeHelper.sol:18]: You mean to simulate a flash-loan behavior?

**Comment by itirabasso** [comment]
I'm closing this PR in favor of #33

**Pull request closed without merging**